### PR TITLE
[SOAR-3792] - InsightIDR Advanced Query

### PR DESCRIPTION
**Pull request by jmcadams-r7**
## Proposed Changes

This PR adds the Advanced Query action to the InsightIDR plugin

## Validation: 
```
RMT-MBP-5357:rapid7_insightidr jmcadams$ icon-validate .
[*] Validating plugin at .

[*] Running Integration Validators...
[*] Executing validator HelpValidator
[*] Executing validator ChangelogValidator
[*] Executing validator RequiredKeysValidator
[*] Executing validator UseCaseValidator
[*] Executing validator SpecPropertiesValidator
[*] Executing validator SpecVersionValidator
[*] Executing validator FilesValidator
[*] Executing validator TagValidator
[*] Executing validator DescriptionValidator
[*] Executing validator TitleValidator
[*] Executing validator VendorValidator
[*] Executing validator DefaultValueValidator
[*] Executing validator IconValidator
[*] Executing validator RequiredValidator
[*] Executing validator VersionValidator
[*] Executing validator DockerfileParentValidator
[*] Executing validator LoggingValidator
[*] Executing validator ProfanityValidator
[*] Executing validator AcronymValidator
[*] Executing validator JSONValidator
[*] Executing validator OutputValidator
[*] Executing validator RegenerationValidator
[*] Executing validator HelpInputOutputValidator
[*] Executing validator SupportValidator
[*] Executing validator RuntimeValidator
[*] Executing validator VersionPinValidator
[*] Executing validator ExampleInputValidator
[*] Plugin successfully validated!

----
[*] Total time elapsed: 680.3430000000001ms
```

## Testing: 
```
RMT-MBP-5357:rapid7_insightidr jmcadams$ ./run.sh -R tests/advanced_query.json -j
Running: cat tests/advanced_query.json | docker run --rm   -i rapid7/rapid7_insightidr:1.4.0  run | grep -- ^\{ | jq -r '.body | try(.log | split("\n") | .[]),.output'
Connect: Connecting...
rapid7/Rapid7 InsightIDR:1.4.0. Step name: advanced_query
Getting log entries from: https://us.api.insight.rapid7.com/log_search/management/logs
Starting new HTTPS connection (1): us.api.insight.rapid7.com:443
https://us.api.insight.rapid7.com:443 "GET /log_search/management/logs HTTP/1.1" 200 1120
Checking Endpoint Agents against Endpoint Agents
Log found.
Found log with name Endpoint Agents and ID: 64e0f0e9-1a88-4601-ae36-f6d03687f313
Getting logs from: https://us.api.insight.rapid7.com/log_search/query/logs/64e0f0e9-1a88-4601-ae36-f6d03687f313
Using parameters: {'query': 'where(iagent)', 'from': 1130778669000, 'to': 1609275548000}
https://us.api.insight.rapid7.com:443 "GET /log_search/query/logs/64e0f0e9-1a88-4601-ae36-f6d03687f313?query=where%28iagent%29&from=1130778669000&to=1609275548000 HTTP/1.1" 202 303
Got a callback url. Polling results...
Trying to get results from callback URL: https://us.api.insight.rapid7.com/log_search/query/fd61b4b6-dd0e-41bc-98aa-12eee8c91525:0:89bbba6cb4c63507cdeaa814c944ac0da4a16e10:50:0d43b70ab19e428d1d9beb9638721f1b0e7751c2:
https://us.api.insight.rapid7.com:443 "GET /log_search/query/fd61b4b6-dd0e-41bc-98aa-12eee8c91525:0:89bbba6cb4c63507cdeaa814c944ac0da4a16e10:50:0d43b70ab19e428d1d9beb9638721f1b0e7751c2: HTTP/1.1" 200 264
Results were not ready. Sleeping 1 second and trying again.
Time left: 60
https://us.api.insight.rapid7.com:443 "GET /log_search/query/fd61b4b6-dd0e-41bc-98aa-12eee8c91525:0:89bbba6cb4c63507cdeaa814c944ac0da4a16e10:50:0d43b70ab19e428d1d9beb9638721f1b0e7751c2: HTTP/1.1" 200 None
Sending results to orchestrator.
Connect: Connecting...
rapid7/Rapid7 InsightIDR:1.4.0. Step name: advanced_query
Getting log entries from: https://us.api.insight.rapid7.com/log_search/management/logs
Checking Endpoint Agents against Endpoint Agents
Log found.
Found log with name Endpoint Agents and ID: 64e0f0e9-1a88-4601-ae36-f6d03687f313
Getting logs from: https://us.api.insight.rapid7.com/log_search/query/logs/64e0f0e9-1a88-4601-ae36-f6d03687f313
Using parameters: {'query': 'where(iagent)', 'from': 1130778669000, 'to': 1609275548000}
Got a callback url. Polling results...
Trying to get results from callback URL: https://us.api.insight.rapid7.com/log_search/query/fd61b4b6-dd0e-41bc-98aa-12eee8c91525:0:89bbba6cb4c63507cdeaa814c944ac0da4a16e10:50:0d43b70ab19e428d1d9beb9638721f1b0e7751c2:
Results were not ready. Sleeping 1 second and trying again.
Time left: 60
Sending results to orchestrator.

{
  "results": [
    {
      "labels": [],
      "sequence_number": 3237168305197691000,
      "timestamp": 1601598638768,
      "log_id": "64e0f0e9-1a88-4601-ae36-f6d03687f313",
      "message": {
        "timestamp": "2020-10-02T00:29:14.649Z",
        "destination_asset": "iagent-win7",
        "source_asset_address": "10.4.22.107",
        "destination_asset_address": "iagent-win7",
        "destination_local_account": "user",
        "logon_type": "NETWORK",
        "result": "SUCCESS",
        "new_authentication": "false",
        "service": "ntlmssp ",
        "source_json": {
          "sourceName": "Microsoft-Windows-Security-Auditing",
          "insertionStrings": [
            "S-1-0-0",
            "-",
            "-",
            "0x0",
            "S-1-5-21-1663271794-3210079414-3662025226-1000",
            "user",
            "iagent-win7",
            "0x204f163c",
            "3",
            "NtLmSsp ",
            "NTLM",
            "",
            "{00000000-0000-0000-0000-000000000000}",
            "-",
            "NTLM V2",
            "128",
            "0x0",
            "-",
            "10.4.22.107",
            "59090"
          ],
          "eventCode": 4624,
          "computerName": "iagent-win7",
          "sid": "",
          "isDomainController": false,
          "eventData": null,
          "timeWritten": "2020-10-02T00:29:13.670722000Z"
        }
      },
      "links": [
        {
          "rel": "Context",
          "href": "https://us.api.insight.rapid7.com/log_search/query/context/3237168305197690880?per_page=50&timestamp=1601598638768&log_keys=64e0f0e9-1a88-4601-ae36-f6d03687f313&context_type=SURROUND"
        }
      ],
      "sequence_number_str": "3237168305197690880"
    },
    {
      "labels": [],
      "sequence_number": 3237168305197695000,
      "timestamp": 1601598638768,
      "log_id": "64e0f0e9-1a88-4601-ae36-f6d03687f313",
      "message": {
        "timestamp": "2020-10-02T00:29:18.376Z",
        "destination_asset": "iagent-win7",
        "source_asset_address": "10.4.22.107",
        "destination_asset_address": "iagent-win7",
        "destination_local_account": "user",
        "logon_type": "NETWORK",
        "result": "SUCCESS",
        "new_authentication": "false",
        "service": "ntlmssp ",
        "source_json": {
          "sourceName": "Microsoft-Windows-Security-Auditing",
          "insertionStrings": [
            "S-1-0-0",
            "-",
            "-",
            "0x0",
            "S-1-5-21-1663271794-3210079414-3662025226-1000",
            "user",
            "iagent-win7",
            "0x204f24b8",
            "3",
            "NtLmSsp ",
            "NTLM",
            "LAB-DEMO-ENGINE",
            "{00000000-0000-0000-0000-000000000000}",
            "-",
            "NTLM V2",
            "128",
            "0x0",
            "-",
            "10.4.22.107",
            "52882"
          ],
.
.
.
<omitted for size>
```

### CODE REVIEW

File: rapid7_insightidr/komand_rapid7_insightidr/actions/advanced_query/action.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+                response.raise_for_status()
+            except Exception:
+                raise PluginException(cause="Failed to get logs from InsightIDR",
+                                      assistance=f"Could not get logs from: {callback_url}",
+                                      data=response.text)
+
+            results_object = response.json()
+            log_entries = results_object.get("events")
+            counter-=1
+
```
Comment (ncramer-r7) [inline rapid7_insightidr/komand_rapid7_insightidr/actions/advanced_query/action.py:106]: Is it worth it to raise a plugin exception (or just more specific logging) here if `log_entries` is empty by the time counter hits 0?

File: rapid7_insightidr/plugin.spec.yaml [hunk 2]
```diff
@@ -494,3 +495,73 @@ actions:
...
+          - Unparsed Data
+          - Third Party Alert
+          - Virus Alert
+          - Web Access Log
+          - Windows Defender
+          - Web Proxy Activity
+    output:
+      results:
+        title: Query Results
+        description: Qurey Results
```
Comment (ncramer-r7) [inline rapid7_insightidr/plugin.spec.yaml:565]:
```suggestion
        description: Query Results
```

File: rapid7_insightidr/komand_rapid7_insightidr/actions/advanced_query/action.py (hunk 1 above)
Comment (jmcadams-r7) [reply rapid7_insightidr/komand_rapid7_insightidr/actions/advanced_query/action.py:106]: Fixed, moved some code around.

File: rapid7_insightidr/help.md [hunk 3]
```diff
@@ -36,6 +36,102 @@ Example input:
 
 ### Actions
 
+#### Advanced Query
+
+This action is used to realtime query into Insight IDR logs.
```
Comment (mhofert-r7) [inline rapid7_insightidr/help.md:41]: This sentence is awkward. Maybe:

This action is used to run realtime queries into Insight IDR logs.

Comment (jmcadams-r7) [reply rapid7_insightidr/help.md]: Fixed

**Review by mhofert-r7 — APPROVED** [review]

**Review by ncramer-r7 — APPROVED** [review]

**Pull request merged**